services/object-mapping-indexer/src/services/objectMappingRouter/index.ts

```diff
@@ -106,25 +117,40 @@ const unsubscribeRecoverObjectMappings = (subscriptionId: string) => {
   }
 }
 
+// Limits the maximum piece index to the already archived segments
+// avoiding to distribute pieces that are not yet archived
+const getUpperLimit = (pieceIndex: number) => {
+  const segmentIndex = segmentUseCase.getSegmentByPieceIndex(pieceIndex)
+  if (segmentIndex >= state.lastRealtimeSegmentIndex) {
+    const UPPER_LIMIT_RANGE_INDEX = 1
+    return segmentUseCase.getPieceIndexRangeBySegment(
+      state.lastRealtimeSegmentIndex,
+    )[UPPER_LIMIT_RANGE_INDEX]
+  }
+  return pieceIndex
+}
+
 const emitRecoverObjectMappings = async () => {
   const recovering = Array.from(
     state.recoverObjectMappingsSubscriptions.entries(),
   )
 
   const promises = recovering.map(
     async ([subscriptionId, { connection, pieceIndex, step }]) => {
-      const result = await objectMappingUseCase.getObjectByPieceIndexAndStep(
+      const upperLimit = getUpperLimit(pieceIndex + step)
+      const result = await objectMappingUseCase.getObjectByPieceIndexRange(
         pieceIndex,
-        step,
+        upperLimit,
       )
 
       state.recoverObjectMappingsSubscriptions.set(subscriptionId, {
         connection,
-        pieceIndex: pieceIndex + step,
+        pieceIndex: upperLimit,
         step,
       })
 
-      if (pieceIndex >= state.lastRealtimeBlockNumber) {
+      const hasReachedLastSegment = pieceIndex !== upperLimit
```

Comment:
Maybe I'm confused about what is happening here, but I'm not sure if this comparison is correct. What if the upperLimit doesn't have any mappings?

(Since segments consist of pairs of data and parity pieces, this is actually guaranteed to happen, because the last piece in a segment is a parity piece with no mappings.)

```suggestion
      const hasReachedLastSegment = pieceIndex > upperLimit
```

Reply:
There was a typo on this one the correct comparison is: `pieceIndex + step != upperLimit` or `pieceIndex + step > upperLimit`.

Reply:
Being `upperPieceIndex = pieceIndex + step`, `upperPieceIndex > upperLimit` and `upperPieceIndex !== upperLimit` are both correct.

Since `upperLimit` is effectively `upperLimit = Math.min(upperPieceIndex, lastPieceFromLastArchivedSegment) `

If `upperPieceIndex == upperLimit` is basically because `upperPieceIndex <= lastPieceFromLastArchivedSegment` so we didn't reach the current archiving segment. 
If `upperPieceIndex > upperLimit` means that `upperPieceIndex <= lastPieceFromLastArchivedSegment` it's just because `upperPieceIndex` is already within the current archiving segment.
The state `upperPieceIndex < upperLimit` is not posible because by definition `Math.min(upperPieceIndex, lastPieceFromLastArchivedSegment)` it cannot be, thus `!==` or `>` are effectively the same.

Anyways, I agree that is clearer to use `>`.